keras/utils/data_utils.py

```diff
@@ -55,23 +57,93 @@ def chunk_read(response, chunk_size=8192, reporthook=None):
     from six.moves.urllib.request import urlretrieve
 
 
+def extract_archive(file_path, path='.', archive_format='auto'):
+    """Extracts an archive if it matches the tar, tar.gz, tar.bz, or zip formats
+
+    # Arguments
+        file_path: path to the archive file
+        path: path to extract the archive file
+        archive_format: Archive format to try for extracting the file.
+                        Options are 'auto', 'tar', 'zip', and None.
+                        'tar' includes tar, tar.gz, and tar.bz files.
+                        The default 'auto' is ['tar', 'zip'].
+                        None or an empty list will return no matches found.
+
+    # Return:
+        True if a match was found and an archive extraction was completed,
+        False otherwise.
+    """
+    if archive_format is None:
+        return False
+    if archive_format is 'auto':
+        archive_format = ['tar', 'zip']
+    if isinstance(archive_format, six.string_types):
+        archive_format = [archive_format]
+
+    for archive_type in archive_format:
+        if archive_type is 'tar':
+            open_fn = tarfile.open
+            is_match_fn = tarfile.is_tarfile
+        if archive_type is 'zip':
+            open_fn = zipfile.ZipFile
+            is_match_fn = zipfile.is_zipfile
+
+        if is_match_fn(file_path):
+            with open_fn(file_path) as archive:
+                try:
+                    archive.extractall(path)
+                except (Exception, KeyboardInterrupt) as e:
+                    if os.path.exists(path):
+                        if os.path.isfile(path):
+                            os.remove(path)
+                        else:
+                            shutil.rmtree(path)
+                    raise
+            return True
+    return False
+
+
 def get_file(fname, origin, untar=False,
-             md5_hash=None, cache_subdir='datasets'):
+             md5_hash=None, cache_subdir='datasets',
+             file_hash=None,
+             hash_algorithm='auto',
+             extract=False,
+             archive_format='auto'):
     """Downloads a file from a URL if it not already in the cache.
 
-    Passing the MD5 hash will verify the file after download
-    as well as if it is already present in the cache.
+    Passing the hash will verify the file after download which also
+    means it will skip re-downloading if it is already present in the cache.
```

Comment:
I corrected the description to match the code, the current master re-downloads the file every time if the path provided is not in the Keras database. Perhaps this is because:

```python
    datadir_base='~/.keras'
    cache_subdir='/path/to/new/destination'
    datadir = os.path.join(datadir_base, cache_subdir)
    print datadir
    # datadir is '/path/to/new/destination'
```

I thought this was clever and good, but perhaps not? If not, I still believe it would be worthwhile to have a way to specify absolute path destinations for all downloaded/extracted files.